ast/term.go

```diff
@@ -1128,52 +1134,73 @@ func ArrayTerm(a ...*Term) *Term {
 	return NewTerm(NewArray(a...))
 }
 
+// Array represents an array as defined by the language. Arrays are similar to the
+// same types as defined by JSON with the exception that they can contain Vars
+// and References.
+type Array interface {
+	Value
+	Copy() Array
+	Equal(other Value) bool
+	Get(pos *Term) *Term
+	Sorted() Array
+	Len() int
+	Elem(i int) *Term
+	Slice(i, j int) Array
+	Until(f func(*Term) bool) bool
+	Iter(f func(*Term) error) error
+	Foreach(f func(*Term))
+	Append(v *Term) Array
+	set(i int, v *Term)
+	elems() []*Term
+}
+
 // NewArray creates an Array with the terms provided. The array will
 // use the provided term slice.
-func NewArray(a ...*Term) *Array {
+func NewArray(a ...*Term) Array {
```

Comment:
Hrrm that change has some potential for breaking existing code on an OPA upgrade.

Reply:
Yes, this is a breaking change.
The benefit of this approach is that we cover more cases, e.g. when checking the length, fetching a specific entry, or doing an actual comparison `data.a == data.b` and `a` and `b` might have different types. These might be edge cases of the more extreme kind, however, so perhaps solving this in `topdown`, as you suggest, is a preferable alternative to a breaking change 🤔 .
Instead of unifying at 

https://github.com/open-policy-agent/opa/blob/00ec6348a41e010590acf1d752e2ff7ad393440d/topdown/eval.go#L376C11-L376C11

it's probably possible to check if the term is truthy. Unification does quite a lot of lifting here, though, so hard to say how much work that entails.

Reply:
Yeah that's along the lines that I thought we _might_ be able to approach this. FWIW `rterm` is fresh, so that simplifies things. I think the case we're interested in optimizing would eventually hit [`biunifyRef`](https://github.com/open-policy-agent/opa/blob/00ec6348a41e010590acf1d752e2ff7ad393440d/topdown/eval.go#L1025); and through `(evalTree).enumerate()` do [`Resolve` here](https://github.com/open-policy-agent/opa/blob/00ec6348a41e010590acf1d752e2ff7ad393440d/topdown/eval.go#L2116)...